exchanges/dydx/dydx.go

```diff
@@ -0,0 +1,963 @@
+func (dy *DYDX) GetLiquidityRewards(ctx context.Context, epoch int64, secondaryAddress string) (*LiquidityProviderRewards, error) {
+	params := url.Values{}
+	if epoch != 0 {
+		params.Set("epoch", strconv.FormatInt(epoch, 10))
+	}
+	if secondaryAddress != "" {
+		params.Set("secondaryAddress", secondaryAddress)
+	}
+	var resp *LiquidityProviderRewards
+	return resp, dy.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, defaultV3EPL, http.MethodGet, common.EncodeURLValues(liquidityRewards, params), nil, &resp)
+}
+
+// GetRetroactiveMiningRewards retrieves the retroactive mining rewards of a given epoch.
+func (dy *DYDX) GetRetroactiveMiningRewards(ctx context.Context) (*RetroactiveMining, error) {
+	var resp *RetroactiveMining
+	return resp, dy.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, defaultV3EPL, http.MethodGet, rewardsRetroactiveMining, nil, &resp)
+}
+
+// SendVerificationEmail send an email to the email address associated with the user, requesting that they click on a link to verify their email address.
+func (dy *DYDX) SendVerificationEmail(ctx context.Context) (interface{}, error) {
+	var resp interface{}
+	return resp, dy.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, sendVerificationEmailEPL, http.MethodGet, emailsSendVeroficationEmail, nil, &resp)
+}
+
+// RequestTestnetTokens requests tokens on dYdX's staging server.
+// a fixed number of tokens will be transferred to the account. Please take note of rate limits.
+func (dy *DYDX) RequestTestnetTokens(ctx context.Context) (*TestnetToken, error) {
+	var resp *TestnetToken
+	return resp, dy.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, defaultV3EPL, http.MethodPost, testnetTokens, nil, &resp)
+}
+
+// GetPrivateProfile retrieves private profile data for the user. This is a superset of the /v3/profile/:publicId endpoint.
+func (dy *DYDX) GetPrivateProfile(ctx context.Context) (*PrivateProfile, error) {
+	var resp *PrivateProfile
+	return resp, dy.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, defaultV3EPL, http.MethodGet, privateProfile, nil, &resp)
+}
+
+// SendAuthenticatedHTTPRequest sends an authenticated HTTP request
+func (dy *DYDX) SendAuthenticatedHTTPRequest(ctx context.Context, endpoint exchange.URL, epl request.EndpointLimit, method, path string, data, result interface{}) error {
+	urlPath, err := dy.API.Endpoints.GetURL(endpoint)
+	if err != nil {
+		return err
+	}
+	var dataString string
+	if data == nil {
+		dataString = ""
+	} else {
+		var value []byte
+		value, err = json.Marshal(data)
+		if err != nil {
+			return err
+		}
+		dataString = string(value)
+	}
+	var creds *account.Credentials
+	creds, err = dy.GetCredentials(ctx)
+	if err != nil {
+		return err
+	}
+	newRequest := func() (*request.Item, error) {
+		var body io.Reader
+		var payload []byte
+		if data != nil {
+			payload, err = json.Marshal(data)
+			if err != nil {
+				return nil, err
+			}
+			body = bytes.NewBuffer(payload)
+		}
```

Comment:
This doesn't seem to have a timestamp or other variable that changes per request so could probably do this outside of newRequest function.